charmcraft/models/project.py

```diff
@@ -323,6 +323,15 @@ class CharmcraftBuildPlanner(models.BuildPlanner):
     build_base: str | None = None
     platforms: dict[str, models.Platform | None] | None = None  # type: ignore[assignment]
 
+    @override
+    @pydantic.field_validator("platforms", mode="before")
+    @classmethod
+    def _populate_platforms(cls, platforms: dict[str, Any]) -> dict[str, Any]:
+        """Overrides the validator to prevent platforms from being modified.
+
+        Modifying the platforms field can break multi-base builds."""
```

Comment:
Is this related to https://github.com/canonical/craft-platforms/pull/84? If so, can we undo this?

Reply:
I don't think this can be avoided. Now that the shorthand and longhand notation are not equivalent, craft-application's expansion of shorthand into longhand will break in one way or another.

If you write:
```
platforms:
  ubuntu@24.04:riscv64:
```

This validator in craft-application could expand it to:
```
platforms:
  ubuntu@24.04:riscv64:
    build-on: [riscv64]
    build-for: [riscv64]
```
but this is a problem because craft-platforms does not allow for a base in platform name and `build-on` or `build-for` entries.

Craft-application could expand it to
```
platforms:
  riscv64:
    build-on: [ubuntu@24.04:riscv64]
    build-for: [ubuntu@24.04:riscv64]
```
but this is a problem for 2 reasons: the platform name has changed, which is confusing when using `--platform` and the build plan could contain duplicate platform names.

Reply:
The problem is that you can't have a base in the platform name and `build-on` or `build-for` entries. At least, that is how it is implemented in craft-platforms.  This is one of the discussion items in the spec that is not clearly defined.

Reply:
Yeah, I wish I would have clarified this earlier as we were all interpreting the spec differently. 

[This is already in](https://github.com/canonical/craft-platforms/blob/7680af231de67febeb6e97d1f572339140c7b0a9/craft_platforms/charm/_build.py#L65-L72) craft-platforms, @cmatsuoka is OK with keeping as-is, and it's more restrictive, I think we should remain on this course.  In a subsequent iteration, we can relax the constraint in craft-platforms and clarify the spec.

Reply:
I see the problem there. The problem I see is that we're not expanding the shorthand notation according to the shorthand notation rules, which say that the platform name should be the same as `build-on` and `build-for` values. I think we'll need to check if there's a shorthand name and `build-on`/`build-for` specified only on entries that have not been expanded, otherwise this will always fail.

Reply:
So
```
plaforms:
  ubuntu@24.04:amd64:
```
would expand to
```
platforms:
  ubuntu@24.04:amd64:
    build-on: [ubuntu@24.04:amd64]
    build-for: [ubuntu@24.04:amd64]
```
but this entry wouldn't be checked again for presence of shorthand name + attributes (because we already know that this will be exactly the case).

Reply:
Maybe the easiest way to do that is to check for shorthand name + attributes *before* doing the expansion.

Reply:
As discussed with @mr-cal, preventing Charmcraft from changing platform entries and sending the input as-is to craft-platforms seems to be the right thing to do: everything related to platforms would be handled by craft-platforms, and this would allow consistent platforms processing for anyone using the library (including Launchpad in the future).

Reply:
I think we're on the same page now, though we should probably work upstream on craft-application not changing the platforms entries.